iot-hub/Samples/service/IotHubSasCredentialAuthenticationSample/Program.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using Azure;
+using Azure.Core;
+using CommandLine;
+using Microsoft.Azure.Devices;
+using System;
+using System.Globalization;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace IotHubSasCredentialAuthenticationSample
+{
+    public class Program
+    {
+        /// <summary>
+        /// A sample to illustrate how to use SAS token for authentication to the IoT hub.
+        /// <param name="args">Run with `--help` to see a list of required and optional parameters.</param>
+        public static async Task Main(string[] args)
+        {            
+            // Parse application parameters
+            Parameters parameters = null;
+            ParserResult<Parameters> result = Parser.Default.ParseArguments<Parameters>(args)
+                .WithParsed(parsedParams =>
+                {
+                    parameters = parsedParams;
+                })
+                .WithNotParsed(errors =>
+                {
+                    Environment.Exit(1);
+                });
+
+            // Initialize SAS token credentials.
+            Console.WriteLine("Creating sas credential.");
+
+            TimeSpan timeToLive = TimeSpan.FromHours(1);
+            DateTime expiresOn = DateTime.UtcNow.Add(timeToLive);
+            string sasToken = GenerateSasToken(parameters.HostName, parameters.SharedAccessKey, expiresOn);
+            // Note: Pass the generated sasToken and not just the shared access signature when creating the AzureSasCredential.
+            AzureSasCredential sasCredential = new AzureSasCredential(sasToken);
+
+            // This is how the credential can be updated in the IotHubSasCredential object whenever necessary.
+            // This sample just shows how to perform the update but it is not necessary to update the token
+            // until the token is close to its expiry.
+            // NOTE: Ensure to pass the new expiry date time.
+            DateTime newExpiresOn = DateTime.UtcNow.Add(timeToLive);
+            string updatedSasToken = GenerateSasToken(parameters.HostName, parameters.SharedAccessKey, expiresOn);
+            sasCredential.Update(updatedSasToken);
+
+            // There are constructors for all the other clients where you can pass SAS credentials - JobClient, RegistryManager, DigitalTwinClient
+            using var serviceClient = ServiceClient.Create(parameters.HostName, sasCredential, parameters.TransportType);
+
+            var sample = new IotHubSasCredentialAuthenticationSample();
+            await sample.RunSampleAsync(serviceClient, parameters.DeviceId);
+        }
+
+        private static string GenerateSasToken(string hostName, string sharedAccessKey, DateTime expiresOn, string policyName = default )
```

Comment:
In the Java sample, I made this function allow for security scoping. Instead of asking for the hostname, I asked for the resource URI that the SAS token should grant access to.

Reply:
```java
@param resourceUri The resource that the shared access token should grant access to. For cases where the token
 * will be used for more than one function (i.e. used by registryManager to create a device and used by serviceClient
 * to send cloud to device messages), this value should be the hostName of your IoT Hub
 * ("my-azure-iot-hub.azure-devices.net" for example). Shared access signatures do support scoping of the resource
 * authorization by making this resourceUri more specific. For example, a resourceUri of "my-azure-iot-hub.azure-devices.net/devices"
 * will make this token only usable when creating/updating/deleting device identities.
 * For more examples, see <a href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#use-security-tokens-from-service-components">Using security tokens from service components</a>
```

Reply:
I think a customer who is interested in generating SAS tokens would be interested in this feature, so consider leaving a note about this scoping option